From ee837ab2b506f6fa40baa65f0fe5e88f47d74f36 Mon Sep 17 00:00:00 2001
From: Brendan King <Brendan.King@imgtec.com>
Date: Fri, 19 Nov 2021 16:57:53 +0000
Subject: [PATCH 1/9] glamor: improve OpenGLES support

Use an OpenGLES 3.0 context, where possible.

OpenGLES 3.0 has map buffer range, so use it when preparing a pixmap for
CPU access.

OpenGLES may support GL_EXT_buffer_storage.

No output was being produced for the zero line width circle and
ellipse tests in x11perf, due to no points being drawn by
glamor_poly_point_gl. This was due to gl_PointSize not being set
in the poly_point vertex shader, something that is not required
for desktop OpenGL.

Commit:

    commit dbfbe58b94ec07a45fd6170f96bffec55051724e
    Author: Corentin Rossignon <corentin.rossignon@gmail.com>
    Date:   Thu Jan 4 08:34:01 2018 +0100
    glamor: Specify GLSL version for xv shader

broke the Xv shader on OpenGLES, by adding a "#version 120"
statement to the shader source. The original problem was the use
of "sample" as a variable name. Remove the #version statement, and
change the variable name from "sample" to "smp".

A later commit:

    commit dcba460af3eedb9d41986bd65f4502998b7a5a6c
    Author: Konstantin <ria.freelander@gmail.com>
    Date:   Tue Jun 28 12:28:39 2022 +0300
    glamor: fix XVideo run with GLES

Worked around the Xv shader issue by forcing the GLSL version to 100 for
GLES. Revert the commit, as it breaks the GLES GLSL version 300 shaders
added by a later IMG commit.
---
 glamor/glamor.c         | 14 ++++++++------
 glamor/glamor_egl.c     | 24 +++++++++++++++++++++++-
 glamor/glamor_points.c  |  3 ++-
 glamor/glamor_prepare.c | 25 +++++++++++++++++++------
 glamor/glamor_program.c |  5 -----
 glamor/glamor_xv.c      | 16 +++++++---------
 6 files changed, 59 insertions(+), 28 deletions(-)

diff --git a/glamor/glamor.c b/glamor/glamor.c
index 848b484cf..df8005c00 100644
--- a/glamor/glamor.c
+++ b/glamor/glamor.c
@@ -762,10 +762,6 @@ glamor_init(ScreenPtr screen, unsigned int flags)
         epoxy_has_gl_extension("GL_ARB_instanced_arrays"))
         glamor_priv->use_gpu_shader4 = epoxy_has_gl_extension("GL_EXT_gpu_shader4");
 
-    glamor_priv->has_rw_pbo = FALSE;
-    if (!glamor_priv->is_gles)
-        glamor_priv->has_rw_pbo = TRUE;
-
     glamor_priv->has_khr_debug = epoxy_has_gl_extension("GL_KHR_debug");
     glamor_priv->has_pack_invert =
         epoxy_has_gl_extension("GL_MESA_pack_invert");
@@ -773,9 +769,11 @@ glamor_init(ScreenPtr screen, unsigned int flags)
         epoxy_has_gl_extension("GL_EXT_framebuffer_blit");
     glamor_priv->has_map_buffer_range =
         epoxy_has_gl_extension("GL_ARB_map_buffer_range") ||
-        epoxy_has_gl_extension("GL_EXT_map_buffer_range");
+        epoxy_has_gl_extension("GL_EXT_map_buffer_range") ||
+        (glamor_priv->is_gles && gl_version >= 30);
     glamor_priv->has_buffer_storage =
-        epoxy_has_gl_extension("GL_ARB_buffer_storage");
+        epoxy_has_gl_extension("GL_ARB_buffer_storage") ||
+        epoxy_has_gl_extension("GL_EXT_buffer_storage");
     glamor_priv->has_mesa_tile_raster_order =
         epoxy_has_gl_extension("GL_MESA_tile_raster_order");
     glamor_priv->has_nv_texture_barrier =
@@ -804,6 +802,10 @@ glamor_init(ScreenPtr screen, unsigned int flags)
 
     glamor_priv->can_copyplane = (gl_version >= 30);
 
+    glamor_priv->has_rw_pbo = FALSE;
+    if (!glamor_priv->is_gles || glamor_priv->has_map_buffer_range)
+        glamor_priv->has_rw_pbo = TRUE;
+
     glamor_setup_debug_output(screen);
 
     glamor_priv->use_quads = !glamor_priv->is_gles &&
diff --git a/glamor/glamor_egl.c b/glamor/glamor_egl.c
index c35b10d83..6a30c52a4 100644
--- a/glamor/glamor_egl.c
+++ b/glamor/glamor_egl.c
@@ -1045,6 +1045,28 @@ glamor_egl_init(ScrnInfoPtr scrn, int fd)
         }
     }
 
+    if (glamor_egl->context == EGL_NO_CONTEXT) {
+        static const EGLint config_attribs[] = {
+            EGL_CONTEXT_CLIENT_VERSION, 3,
+            EGL_CONTEXT_MINOR_VERSION_KHR, 0,
+            EGL_NONE
+        };
+        if (eglBindAPI(EGL_OPENGL_ES_API)) {
+            glamor_egl->context = eglCreateContext(glamor_egl->display,
+                                                   NULL, EGL_NO_CONTEXT,
+                                                   config_attribs);
+        }
+
+        if (glamor_egl->context != EGL_NO_CONTEXT) {
+            if (!eglMakeCurrent(glamor_egl->display,
+                                EGL_NO_SURFACE, EGL_NO_SURFACE, glamor_egl->context)) {
+                xf86DrvMsg(scrn->scrnIndex, X_ERROR,
+                           "Failed to make GLES3 context current\n");
+                goto error;
+            }
+        }
+    }
+
     if (glamor_egl->context == EGL_NO_CONTEXT) {
         static const EGLint config_attribs[] = {
             EGL_CONTEXT_CLIENT_VERSION, 2,
@@ -1062,7 +1084,7 @@ glamor_egl_init(ScrnInfoPtr scrn, int fd)
 
         if (glamor_egl->context == EGL_NO_CONTEXT) {
             xf86DrvMsg(scrn->scrnIndex, X_ERROR,
-                       "glamor: Failed to create GL or GLES2 contexts\n");
+                       "glamor: Failed to create GL or GLES contexts\n");
             goto error;
         }
 
diff --git a/glamor/glamor_points.c b/glamor/glamor_points.c
index 8f0e4b1da..5d371bc45 100644
--- a/glamor/glamor_points.c
+++ b/glamor/glamor_points.c
@@ -32,7 +32,8 @@
 static const glamor_facet glamor_facet_point = {
     .name = "poly_point",
     .vs_vars = "attribute vec2 primitive;\n",
-    .vs_exec = GLAMOR_POS(gl_Position, primitive),
+    .vs_exec = GLAMOR_POS(gl_Position, primitive)
+               "        gl_PointSize = 1.0;\n",
 };
 
 static Bool
diff --git a/glamor/glamor_prepare.c b/glamor/glamor_prepare.c
index fba875e28..5aea1bd18 100644
--- a/glamor/glamor_prepare.c
+++ b/glamor/glamor_prepare.c
@@ -132,12 +132,25 @@ glamor_prep_drawable_box(DrawablePtr drawable, glamor_access_t access, BoxPtr bo
     RegionUninit(&region);
 
     if (priv->pbo) {
-        if (priv->map_access == GLAMOR_ACCESS_RW)
-            gl_access = GL_READ_WRITE;
-        else
-            gl_access = GL_READ_ONLY;
-
-        pixmap->devPrivate.ptr = glMapBuffer(GL_PIXEL_PACK_BUFFER, gl_access);
+        if (glamor_priv->has_map_buffer_range) {
+            if (priv->map_access == GLAMOR_ACCESS_RW)
+                gl_access = GL_MAP_WRITE_BIT;
+            else
+                gl_access = GL_MAP_READ_BIT;
+
+            pixmap->devPrivate.ptr = glMapBufferRange(GL_PIXEL_PACK_BUFFER, 0,
+                                                      pixmap->devKind *
+                                                      pixmap->drawable.height,
+                                                      gl_access);
+        } else {
+            if (priv->map_access == GLAMOR_ACCESS_RW)
+                gl_access = GL_READ_WRITE;
+            else
+                gl_access = GL_READ_ONLY;
+
+            pixmap->devPrivate.ptr = glMapBuffer(GL_PIXEL_PACK_BUFFER,
+                                                 gl_access);
+        }
         glBindBuffer(GL_PIXEL_PACK_BUFFER, 0);
     }
 
diff --git a/glamor/glamor_program.c b/glamor/glamor_program.c
index 9ee945d11..0a8854096 100644
--- a/glamor/glamor_program.c
+++ b/glamor/glamor_program.c
@@ -283,11 +283,6 @@ glamor_build_program(ScreenPtr          screen,
             gpu_shader4 = TRUE;
         }
     }
-    /* For now, fix shader version to GLES as 100. We will fall with 130 shaders
-     * in previous check due to forcibly set 120 glsl for GLES. But this patch
-     * makes xv shaders to work */
-    if(version && glamor_priv->is_gles)
-        version = 100;
 
     vs_vars = vs_location_vars(locations);
     fs_vars = fs_location_vars(locations);
diff --git a/glamor/glamor_xv.c b/glamor/glamor_xv.c
index 59fca909b..edad5a2d7 100644
--- a/glamor/glamor_xv.c
+++ b/glamor/glamor_xv.c
@@ -94,8 +94,6 @@ static const glamor_facet glamor_facet_xv_planar_2 = {
 static const glamor_facet glamor_facet_xv_planar_3 = {
     .name = "xv_planar_3",
 
-    .version = 120,
-
     .source_name = "v_texcoord0",
     .vs_vars = ("attribute vec2 position;\n"
                 "attribute vec2 v_texcoord0;\n"
@@ -111,14 +109,14 @@ static const glamor_facet glamor_facet_xv_planar_3 = {
                 "uniform vec4 vco;\n"
                 "varying vec2 tcs;\n"),
     .fs_exec = (
-                "        float sample;\n"
+                "        float smp;\n"
                 "        vec4 temp1;\n"
-                "        sample = texture2D(y_sampler, tcs).w;\n"
-                "        temp1.xyz = offsetyco.www * vec3(sample) + offsetyco.xyz;\n"
-                "        sample = texture2D(u_sampler, tcs).w;\n"
-                "        temp1.xyz = ucogamma.xyz * vec3(sample) + temp1.xyz;\n"
-                "        sample = texture2D(v_sampler, tcs).w;\n"
-                "        temp1.xyz = clamp(vco.xyz * vec3(sample) + temp1.xyz, 0.0, 1.0);\n"
+                "        smp = texture2D(y_sampler, tcs).w;\n"
+                "        temp1.xyz = offsetyco.www * vec3(smp) + offsetyco.xyz;\n"
+                "        smp = texture2D(u_sampler, tcs).w;\n"
+                "        temp1.xyz = ucogamma.xyz * vec3(smp) + temp1.xyz;\n"
+                "        smp = texture2D(v_sampler, tcs).w;\n"
+                "        temp1.xyz = clamp(vco.xyz * vec3(smp) + temp1.xyz, 0.0, 1.0);\n"
                 "        temp1.w = 1.0;\n"
                 "        gl_FragColor = temp1;\n"
                 ),
-- 
2.34.1

